### ISSUE TITLE
"The Object that can be named is not the real Object"

### ISSUE THREAD

**Issue by masak** [masak/007#179]
5b491c284976320ed41ee9fd8b4d4739f562bc49 just landed, which means this is now a bug.

    $ FLAG_007_CLASS=1 bin/007 -e='class Object {}; say(type(new {}) == type(new Object {}))'
    True

The problem is that the *default* type (which `new {}` uses) is looked up like this:

    my $type = ~($<identifier> // "Object");
    my $type-obj = $*runtime.get-var($type).type;

Oops. That even looks wrong.

Now, `get-var` has an optional second parameter `$frame`, which should be enough to fix this.

Needs a test, of course.

**Comment by vendethiel** [comment]
> Now, get-var has an optional second parameter $frame, which should be enough to fix this.

huh? We only want to use the frame if it's an object. Not sure it even makes sense to use the same parsing rule, to be quite honest.

**Comment by masak** [comment]
> huh? We only want to use the frame if it's an object. Not sure it even makes sense to use the same parsing rule, to be quite honest.

You have a point.

In fact, https://github.com/masak/007/issues/144#issuecomment-236031462 argues for the removal of the `new { ... }` parsing case. That sounded like a good way to fix this, so I just ran the test suite against the following change:

```diff
diff --git a/lib/_007/Parser/Syntax.pm b/lib/_007/Parser/Syntax.pm
index 701a7de..f6fa19b 100644
--- a/lib/_007/Parser/Syntax.pm
+++ b/lib/_007/Parser/Syntax.pm
@@ -223,7 +223,7 @@ grammar _007::Parser::Syntax {
     }
     token term:object {
         [new» <.ws>
-            [<identifier> <?{ $*runtime.maybe-get-var(~$<identifier>) ~~ Val::Type }> <.ws>]?
+            <identifier> <?{ $*runtime.maybe-get-var(~$<identifier>) ~~ Val::Type }> <.ws>
         ]?
         '{' ~ '}' <propertylist>
     }
```

It failed *one* test:

    {
        my $program = q:to/./;
            my o1 = { name: "James" };
            my o2 = new { name: "James" };

            say(o1 == o2);
            .

        outputs
            $program,
            qq[True\n],
            "`new` on object literals without the type is allowed but optional";
    }

This test can easily be removed, since all it's doing is testing the `new { ... }` syntax. Since removing both the code path and the test was so easy, I just made a commit of that.

Even better, I think from this starting point I can easily split it up into two rules `new MyType { ... }` and `{ ... }`, which now seems an excellent idea.

**Comment by vendethiel** [comment]
I'm happy with that outcome.

**Comment by masak** [comment]
Re-opening. Just discovered when starting to fix #198 that this one wasn't really fixed either. Probably the tests made it seem that way because I wasn't running them with `FLAG_007_CLASS` turned on.

This is a real risk, something I'd like to fix on the Travis level (but I don't know how). I fixed it somewhat locally by having my shorthand alias run the tests with our current set of feature flags switched on.
